File: sdk/js/registry.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import { loadRegistry, validate, handshakes, partners, byTag, getEntry, tagCounts } from "./registry.mjs";

const REG_PATH = new URL("../../data/registry.json", import.meta.url).pathname;

test("loads and validates the checked-in registry", async () => {
  const reg = await loadRegistry(REG_PATH);
  assert.equal(reg.title, "arc_registry");
  assert.equal(validate(reg).length, 0);
});

test("handshake count matches the declared total", async () => {
  const reg = await loadRegistry(REG_PATH);
  assert.equal(handshakes(reg).length, reg.total_handshakes);
});

test("partners and projects partition sensibly", async () => {
  const reg = await loadRegistry(REG_PATH);
  assert.equal(partners(reg).length, 8);
  assert.ok(getEntry(reg, "askjimmy").tags.includes("forge"));
});

test("byTag filters correctly", async () => {
  const reg = await loadRegistry(REG_PATH);
  const defi = byTag(reg, "defi").map((e) => e.id);
  assert.deepEqual(defi.sort(), ["askjimmy", "listen-rs"]);
});

test("tagCounts is a descending histogram", async () => {
  const reg = await loadRegistry(REG_PATH);
  const counts = tagCounts(reg);
  for (let i = 1; i < counts.length; i++) assert.ok(counts[i - 1][1] >= counts[i][1]);
});

test("validate flags broken registries", () => {
  const bad = { title: "x", total_handshakes: 2, tags: ["a"], entries: [
    { id: "One", type: "mystery", tags: ["zzz"], summary: "" },
    { id: "One", type: "handshake", tags: [], summary: "" },
  ]};
  const problems = validate(bad);
  assert.ok(problems.some((p) => p.includes("duplicate id")));
  assert.ok(problems.some((p) => p.includes("kebab-case")));
  assert.ok(problems.some((p) => p.includes("unknown tag")));
  assert.ok(problems.some((p) => p.includes("unknown type")));
  assert.ok(problems.some((p) => p.includes("total_handshakes")));
});

File: tests/registry_test.rs
use arc_registry::{EntryType, Registry};

fn load() -> Registry {
    Registry::load(concat!(env!("CARGO_MANIFEST_DIR"), "/data/registry.json")).expect("registry loads")
}

#[test]
fn loads_and_validates() {
    let reg = load();
    assert_eq!(reg.title, "arc_registry");
    assert!(!reg.entries.is_empty());
}

#[test]
fn handshake_count_matches_declared_total() {
    let reg = load();
    assert_eq!(reg.handshakes().count() as u32, reg.total_handshakes);
}

#[test]
fn every_tag_is_in_taxonomy() {
    let reg = load();
    for e in &reg.entries {
        for t in &e.tags {
            assert!(reg.tags.contains(t), "{} uses unknown tag {t}", e.id);
        }
    }
}

#[test]
fn ids_are_unique_and_kebab() {
    let reg = load();
    let mut seen = std::collections::HashSet::new();
    for e in &reg.entries {
        assert!(seen.insert(&e.id), "duplicate id {}", e.id);
        assert!(
            e.id.chars().all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '-'),
            "id {} is not kebab-case",
            e.id
        );
    }
}

#[test]
fn askjimmy_is_the_forge_launch() {
    let reg = load();
    let aj = reg.get("askjimmy").expect("askjimmy present");
    assert_eq!(aj.entry_type, EntryType::Handshake);
    assert!(aj.tags.iter().any(|t| t == "forge"));
    assert_eq!(aj.token.as_deref(), Some("token_live"));
}

#[test]
fn partners_have_no_token_field_semantics() {
    let reg = load();
    for p in reg.partners() {
        assert!(p.token.is_none(), "partner {} should not declare a token status", p.id);
    }
}

#[test]
fn by_tag_filters() {
    let reg = load();
    let defi: Vec<_> = reg.by_tag("defi").map(|e| e.id.as_str()).collect();
    assert!(defi.contains(&"listen-rs"));
    assert!(defi.contains(&"askjimmy"));
    assert!(!defi.contains(&"ryzome"));
}
